### DISCUSSION TITLE
Dynamic naming of build artifacts based on commit reference

### DISCUSSION THREAD

**Discussion by Jademalo** [community/community#25463] [Actions]
<p>I’ve been staring at the docs for about half an hour, but honestly I’ve got no idea what I’m doing so I figure it’s worth asking for some more direct help.</p>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/Jademalo/SurroundFix/blob/73be46e74fd1748e5a02ba3ea6f5c9f9f884164e/.github/workflows/package.yml" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/Jademalo/SurroundFix/blob/73be46e74fd1748e5a02ba3ea6f5c9f9f884164e/.github/workflows/package.yml" target="_blank" rel="noopener nofollow ugc">Jademalo/SurroundFix/blob/73be46e74fd1748e5a02ba3ea6f5c9f9f884164e/.github/workflows/package.yml</a></h4>
<pre><code class="lang-yml"># description of this workflow, can be anything you want
name: Package with Artifacts

# we need to let GitHub know _when_ we want to release, typically only when we create a new tag.
# this will target only tags, and not all pushes to the master branch.
# this part can be heavily customized to your liking, like targeting only tags that match a certain word,
# other branches or even pullrequests.

# https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#example-using-multiple-events-with-activity-types-or-configuration
# workflows don't show up in the action list until they have been run once.
# if a push to a branch would trigger a workflow, it uses the workflow file in
# the branch or tag that triggered it.
# branches-ignore means that master won't trigger from it's own workflow.
# multiple commits in a single push will only run once.

on:
  push:
    branches-ignore:
      - master
    tags-ignore:
</code></pre>

  This file has been truncated. <a href="https://github.com/Jademalo/SurroundFix/blob/73be46e74fd1748e5a02ba3ea6f5c9f9f884164e/.github/workflows/package.yml" target="_blank" rel="noopener nofollow ugc">show original</a>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>My repo has a fairly simple action - When a commit is made to a branch that isn’t master, run the BigWigs packager and upload the artifacts.</p>
<p>The packager creates two zip files, <code>SurroundFix-v3.6-7-g5e389b5.zip</code> and <code>SurroundFix-v3.6-7-g5e389b5-classic.zip</code>, with the reference part of the file being the output of <code>git describe --always</code>. These are what I currently upload with the script.</p>
<p>The packager runs twice to create each zip, and it also creates a folder, simply called <code>SurroundFix</code>. What I would like to do is upload an artifact after the first run that contains this folder called <code>SurroundFix-v3.6-7-g5e389b5</code>, and an artifact after the second run that contains this folder called <code>SurroundFix-v3.6-7-g5e389b5-classic</code>, with the git reference obviously changing depending on the current commit that triggers the build.</p>
<p>That would leave me with two zip artifacts named as above, each containing a folder called <code>SurroundFix</code>.</p>
<p>How exactly would I go about doing that?</p>
<pre><code class="lang-auto">      - name: Upload Retail Artifact
        uses: actions/upload-artifact@v2
        with:
          name: (folder name + git reference)
          path: .release/*/

      - name: Upload Classic Artifact
        uses: actions/upload-artifact@v2
        with:
          name: (folder name + git reference + `-classic`)
          path: .release/*/
</code></pre>
<p>I know that’s basically what I want to do, but I have absolutely no idea how to do that.</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    <img src="" class="thumbnail" width="" height="">

<h3><a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions" target="_blank" rel="noopener nofollow ugc">Context and expression syntax for GitHub Actions - GitHub Docs</a></h3>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
Based on this page, I assume it’s something to do with <code>${{ github.head_ref }}</code> or something similar, but I’ve got absolutely no idea how to find the name of the folder to append to the start, no idea how to get the git reference, and no idea how to combine the values into a complete name.
<p>If anyone can give me any help with this, I’d hugely appreciate it.<br>
Thank you!</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="Jademalo" data-post="1" data-topic="137367">
<div class="title">
<div class="quote-controls"></div>
 Jademalo:</div>
<blockquote>
<p>Based on this page, I assume it’s something to do with <code>${{ github.head_ref }}</code> or something similar, but I’ve got absolutely no idea how to find the name of the folder to append to the start, no idea how to get the git reference, and no idea how to combine the values into a complete name.</p>
</blockquote>
</aside>
<p>I’d use the <code>github.sha</code> variable, unless you want to shorten it first. Combining it just means using <code>${{ github.sha }}</code> in the pattern, like this:</p>
<pre><code class="lang-yaml">      - name: Upload Retail Artifact
        uses: actions/upload-artifact@v2
        with:
          name: SurroundFix-${{ github.sha }}
          path: .release/*/
</code></pre>
<p>I’m not sure what you mean by finding the folder name, I thought that’s fixed? If you’re talking about the version number you can just store it in an environment variable and use the <code>env</code> context to access it, the same way as with the <code>github</code> context.</p>

**Comment by Jademalo** [comment]
<p>Oh, that’s fairly simple, I thought there would have to be something fancy to combine the string.</p>
<p>The folder name should always be the same, but I’d ideally like to try and make it dynamic as there are a couple of potential future sitautions where it might change.</p>
<pre><code class="lang-auto">    env:
      folder_name: (something that returns the name of the folder in .release/)

      - name: Upload Retail Artifact
        uses: actions/upload-artifact@v2
        with:
          name: ${{ folder_name }}-${{ github.sha }}
          path: .release/*/
</code></pre>
<p>Do I need to be doing something like this to get a variable with the name of the folder?<br>
Honestly I’m totally at sea here, thank you for the answer!</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>Basically yes! Though I assume that the folder name is going to be a bit too dynamic to set it in the workflow or job, so it’s probably best to create it in a step and <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#environment-files">write it to the environment file so it’s available to the following steps</a>:</p>
<pre><code class="lang-yaml">      - name: Set folder name
        run: |
          echo "folder_name=$(some_command_to_find_the_folder)" &gt;&gt; $GITHUB_ENV
      - name: Upload Retail Artifact
        uses: actions/upload-artifact@v2
        with:
          name: ${{ env.folder_name }}-${{ github.sha }}
          path: .release/*/
</code></pre>
<p>The code in the <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsrun"><code>run</code> step</a> is regular Bash. You can use other interpreters or call scripts, the important thing is that the <code>folder_name=...</code> assignment ends up in the file given by the <code>GITHUB_ENV</code> environment variable.</p>

**Comment by Jademalo** [comment]
<p>Awesome, I think I should be able to get it to work with this. I know for certain that there’s a way to get a folder name from bash.</p>
<p>Thanks!</p>
<p>EDIT:<br>
As an edit for future reference, I managed to get this working absolutely perfectly.<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/Jademalo/SurroundFix/actions/runs/304339560" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/386846/181091228-baf4fa31-f6e6-49d0-b125-a5ced291bcaf.png" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/Jademalo/SurroundFix/actions/runs/304339560" target="_blank" rel="noopener nofollow ugc">Jademalo/SurroundFix</a></h3>

<p>A World of Warcraft UI AddOn to fix various UI issues with multiple monitor setups - Jademalo/SurroundFix</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<pre><code class="lang-auto">      # add variables for the name of the folder being added as an artifact
      - name: Set Name Variables
        run: |
          echo "version=$(git describe)" &gt;&gt; $GITHUB_ENV
          cd .release
          echo "folder_name=$(ls -d */ | sed 's:/*$::')" &gt;&gt; $GITHUB_ENV

      # upload the artifact for retail
      - name: Upload Retail Artifact
        uses: actions/upload-artifact@v2
        with:
          name: ${{ env.folder_name }}-${{ env.version }}
          path: |
            .release/*/
            !.release/*.zip
</code></pre>
<p>I found a bit of bash that output the directory name and removed the trailing slash, then realised I could do the same thing to just put git describe into a variable too.<br>
Now the artifact names are exactly as I wanted them, and everything is absolutely perfect.</p>
<p>Thanks again for the help!</p>
